src/lib.rs

```diff
@@ -746,15 +746,15 @@ where
 macro_rules! arith_impl {
     (impl $imp:ident, $method:ident) => {
         forward_all_binop!(impl $imp, $method);
-        // Abstracts the a/b `op` c/d = (a*d `op` b*c) / (b*d) pattern
+        // Abstracts a/b `op` c/d = (a*lcm/b `op` c*lcm/d)/lcm where lcm = lcm(b,d)
         impl<T: Clone + Integer> $imp<Ratio<T>> for Ratio<T> {
             type Output = Ratio<T>;
             #[inline]
             fn $method(self, rhs: Ratio<T>) -> Ratio<T> {
-                Ratio::new(
-                    (self.numer * rhs.denom.clone()).$method(self.denom.clone() * rhs.numer),
-                    self.denom * rhs.denom,
-                )
+                let lcm = self.denom.lcm(&rhs.denom.clone());
+                let lhs_numer = self.numer * (lcm.clone() / self.denom);
+                let rhs_numer = rhs.numer * (lcm.clone() / rhs.denom);
+                Ratio::new(lhs_numer.$method(rhs_numer), lcm)
```

Comment:
It might be worth a special case for equal denominators, then we can just do the operation on the numerators directly, saving the cost of `lcm` and normalization. The equality check should be relatively cheap. Make sure we have tests with and without this case though.

Reply:
As it is now, calculating the LCM of a number with itself in this code incurs the cost of one multiplication (+ constant overhead of gcd, which is very efficient for equal arguments [cool chart](https://upload.wikimedia.org/wikipedia/commons/8/85/Euclidean_Algorithm_Running_Time.svg)). Multiplication can be as bad as O(n^2) for inefficiently implemented types, and will never be better than O(n logn). Because of this, I agree that it makes sense to special case equal denominators.
Also there is already a test case -- in `test_add`, both
```
test(_1, _1, _2);
test(_1_2, _3_2, _2);
```
and in `test_sub`
```
test(_3_2, _1_2, _1);
```
surprisingly, test_rem needed new test cases.

Reply:
I read through the gnu multiple precision source (version 6.1.2) and found that for multiplication, they special case equal operands (squaring), but not equal denominators. I like the idea of special casing equal denominators better, as it seems like it will save execution time more often, and checking equal denominators is cheap.

Reply:
Please avoid the possibility of deriving from GPL code here. IANAL to know for sure what counts, but let's not push it.

Reply:
I would guess this case does not count as GPL derived, since I specifically elected not to do it the way GMP does it. To be safe, I'll stop reading through GMP sources.